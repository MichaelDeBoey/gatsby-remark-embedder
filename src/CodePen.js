import { URL } from 'url';

export const shouldTransform = url => {
  const { host, pathname } = new URL(url);

  return (
    (host === 'codepen.io' || host === 'www.codepen.io') &&
    pathname.includes('/pen/')
  );
};

export const getHTML = url => {
  const iframeUrl = url.replace('/pen/', '/embed/preview/');

  return `<iframe src="${iframeUrl}" style="width:100%; height:300px;"></iframe>`;
};
