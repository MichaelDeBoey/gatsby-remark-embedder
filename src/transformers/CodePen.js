import { URL } from 'url';

export const shouldTransform = url => {
  const { host, pathname } = new URL(url);

  return (
    ['codepen.io', 'www.codepen.io'].includes(host) &&
    pathname.includes('/pen/')
  );
};

export const getHTML = url => {
  const iframeUrl = url.replace('/pen/', '/embed/preview/');

  return `<iframe src="${iframeUrl}" style="width:100%; height:300px;"></iframe>`;
};
