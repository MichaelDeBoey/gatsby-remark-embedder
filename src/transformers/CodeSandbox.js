import { URL } from 'url';

export const shouldTransform = url => {
  const { host, pathname } = new URL(url);

  return (
    ['codesandbox.io', 'www.codesandbox.io'].includes(host) &&
    pathname.includes('/s/')
  );
};

export const getHTML = url => {
  const iframeUrl = url.replace('/s/', '/embed/');

  return `<iframe src="${iframeUrl}" style="width:100%; height:500px; border:0; border-radius: 4px; overflow:hidden;" allow="geolocation; microphone; camera; midi; vr; accelerometer; gyroscope; payment; ambient-light-sensor; encrypted-media; usb" sandbox="allow-modals allow-forms allow-popups allow-scripts allow-same-origin"></iframe>`;
};
