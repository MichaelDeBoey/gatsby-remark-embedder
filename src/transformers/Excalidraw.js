import puppeteer from 'puppeteer';

const getImage = async (url) => {
  const browser = await puppeteer.launch({
    // disable sandbox in production
    args: process.env.NODE_ENV === 'production' ? ['--no-sandbox'] : [],
  });

  const page = await browser.newPage();
  await page.goto(url);
  await page.click('[aria-label=Export]');
  await page.click("[aria-label='Scale 3 x']");

  const frame = await page.mainFrame();
  const result = await frame.evaluate(
    () =>
      new Promise((resolve, reject) => {
        try {
          delete window.chooseFileSystemEntries;
          const reader = new FileReader();
          reader.addEventListener('loadend', () => resolve(reader.result));
          reader.addEventListener('error', () => reject(reader.error));
          URL.createObjectURL = (blob) => reader.readAsText(blob);

          const button = document.querySelector('[aria-label="Export to SVG"]');
          button.click();
        } catch (error) {
          reject(error);
        }
      })
  );

  await browser.close();
  return result;
};

export const shouldTransform = (url) =>
  /^https?:\/\/(www\.)?excalidraw\.com\/#json=/.test(url);

export const getHTML = async (url) => {
  const svg = await getImage(url);
  return `<a style="box-shadow: none" href="${url}">${svg
    .replace('<svg', '<svg role="img" aria-label="Excalidraw drawing"')
    .replace(/\n\s*/g, '')}</a>`;
};
