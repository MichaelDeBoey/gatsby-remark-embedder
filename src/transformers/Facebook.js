export const shouldTransform = (url) => {
  const { host, pathname } = new URL(url);

  return (
    ['facebook.com', 'www.facebook.com'].includes(host) &&
    pathname.includes('/videos/')
  );
};

export const getHTML = (string) => {
  const src = `https://facebook.com/plugins/video.php?href=${string}`;
  return `<iframe src="${src}" width="560" height="315" style="border:none;overflow:hidden" scrolling="no" frameborder="0" allowTransparency="true" allowFullScreen="true"></iframe>`;
};
