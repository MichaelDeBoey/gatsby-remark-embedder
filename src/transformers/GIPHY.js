import { fetchOEmbedData } from './utils';

const isMediaSubDomain = (host) => /^(media([0-9]+)?\.)giphy\.com$/.test(host);

export const shouldTransform = (url) => {
  const { host, pathname } = new URL(url);

  return (
    (['giphy.com', 'www.giphy.com'].includes(host) &&
      pathname.includes('/gifs/')) ||
    (isMediaSubDomain(host) && pathname.includes('/media/'))
  );
};

export const getGIPHYId = (url) => {
  const { host, pathname } = new URL(url);

  if (isMediaSubDomain(host)) {
    return pathname.split('/')[2];
  }

  return pathname.split('-').pop();
};

export const getGIPHYResponsivePadding = ({ height, width }) =>
  Math.round((height / width) * 100);

export const getHTML = (url) =>
  fetchOEmbedData(`https://giphy.com/services/oembed?url=${url}`).then(
    ({ height, width }) => {
      const GIPHYId = getGIPHYId(url);
      const padding = getGIPHYResponsivePadding({ height, width });

      return `<div style="width:100%;height:0;padding-bottom:${padding}%;position:relative;"><iframe src="https://giphy.com/embed/${GIPHYId}" width="100%" height="100%" style="position:absolute" frameborder="0" class="giphy-embed" allowfullscreen></iframe></div>`;
    }
  );
