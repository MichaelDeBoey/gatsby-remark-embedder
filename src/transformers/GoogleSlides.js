export const shouldTransform = (url) => {
  const { host, pathname } = new URL(url);

  return (
    host === 'docs.google.com' ||
    (['www.docs.google.com', 'docs.google.com'].includes(host) &&
      pathname.split('/').length === 6)
  );
};

export const getGoogleSlidesFrameSrc = (urlString) => {
  const { pathname } = new URL(urlString);
  const splitedPathName = pathname.split('/');
  const id = splitedPathName[splitedPathName.length - 2];

  const embedUrl = new URL(
    `https://docs.google.com/presentation/d/e/${id}/embed?start=false&loop=false&delayms=3000`
  );

  return embedUrl.toString();
};

export const getHTML = (url) => {
  const iframeUrl = getGoogleSlidesFrameSrc(url);

  return `<iframe src="${iframeUrl}" frameborder="0" width="960" height="569" allowfullscreen="true" mozallowfullscreen="true" webkitallowfullscreen="true"></iframe>`;
};
