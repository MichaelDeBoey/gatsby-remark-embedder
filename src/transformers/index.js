import * as CodePenTransformer from './CodePen';
import * as CodeSandboxTransformer from './CodeSandbox';
import * as GIPHYTransformer from './GIPHY';
import * as InstagramTransformer from './Instagram';
import * as LichessTransformer from './Lichess';
import * as PinterestTransformer from './Pinterest';
import * as SlidesTransformer from './Slides';
import * as SoundCloudTransformer from './SoundCloud';
import * as SpotifyTransformer from './Spotify';
import * as StreamableTransformer from './Streamable';
import * as TestingPlaygroundTransformer from './TestingPlayground';
import * as TwitchTransformer from './Twitch';
import * as TwitterTransformer from './Twitter';
import * as YouTubeTransformer from './YouTube';

export const defaultTransformers = [
  CodePenTransformer,
  CodeSandboxTransformer,
  GIPHYTransformer,
  InstagramTransformer,
  LichessTransformer,
  PinterestTransformer,
  SlidesTransformer,
  SoundCloudTransformer,
  SpotifyTransformer,
  StreamableTransformer,
  TestingPlaygroundTransformer,
  TwitchTransformer,
  TwitterTransformer,
  YouTubeTransformer,
];
