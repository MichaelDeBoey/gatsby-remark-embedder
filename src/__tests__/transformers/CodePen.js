import cases from 'jest-in-case';

import plugin from '../../';
import { getHTML, shouldTransform } from '../../transformers/CodePen';

import { cache, getMarkdownASTForFile, mdastToHtml } from '../helpers';

cases(
  'url validation',
  ({ url, valid }) => {
    expect(shouldTransform(url)).toBe(valid);
  },
  {
    'non-CodePen url': {
      url: 'https://not-a-codepen-url.com',
      valid: false,
    },
    "non-CodePen url ending with 'codepen.io'": {
      url: 'https://this-is-not-codepen.io',
      valid: false,
    },
    "non-CodePen url ending with 'codepen.io' and having '/embed/'": {
      url: 'https://this-is-not-codepen.io/user/embed/123456',
      valid: false,
    },
    "non-CodePen url ending with 'codepen.io' and having '/pen/'": {
      url: 'https://this-is-not-codepen.io/user/pen/123456',
      valid: false,
    },
    'team page': {
      url: 'https://codepen.io/team/codepen',
      valid: false,
    },
    'user page': {
      url: 'https://codepen.io/MichaelDeBoey',
      valid: false,
    },
    'random page': {
      url: 'https://codepen.io/random-page',
      valid: false,
    },
    'blog url': {
      url: 'https://blog.codepen.io',
      valid: false,
    },
    "blog url with '/embed/'": {
      url: 'https://blog.codepen.io/user/embed/123456',
      valid: false,
    },
    "blog url with '/pen/'": {
      url: 'https://blog.codepen.io/user/pen/123456',
      valid: false,
    },
    'Pen embed url': {
      url: 'https://codepen.io/team/codepen/embed/PNaGbb',
      valid: false,
    },
    'Pen embed url with parameters': {
      url: 'https://codepen.io/team/codepen/embed/PNaGbb?default-tab=js',
      valid: false,
    },
    'Team Pen url': {
      url: 'https://codepen.io/team/codepen/pen/PNaGbb',
      valid: true,
    },
    "Team Pen url having 'www' subdomain": {
      url: 'https://www.codepen.io/team/codepen/pen/PNaGbb',
      valid: true,
    },
    'User Pen url': {
      url: 'https://codepen.io/chriscoyier/pen/owBwKM',
      valid: true,
    },
    "User Pen url having 'www' subdomain": {
      url: 'https://www.codepen.io/chriscoyier/pen/owBwKM',
      valid: true,
    },
  }
);

test('Gets the correct CodePen iframe', () => {
  const html = getHTML('https://codepen.io/team/codepen/pen/PNaGbb');

  expect(html).toMatchInlineSnapshot(
    `<iframe src="https://codepen.io/team/codepen/embed/preview/PNaGbb" style="width:100%; height:300px;"></iframe>`
  );
});

test('Plugin can transform CodePen links', async () => {
  const markdownAST = getMarkdownASTForFile('CodePen');

  const processedAST = await plugin({ cache, markdownAST });

  expect(mdastToHtml(processedAST)).toMatchInlineSnapshot(`
    <p>https://not-a-codepen-url.com</p>
    <p>https://this-is-not-codepen.io</p>
    <p>https://this-is-not-codepen.io/user/embed/123456</p>
    <p>https://this-is-not-codepen.io/user/pen/123456</p>
    <p>https://codepen.io/team/codepen</p>
    <p>https://codepen.io/MichaelDeBoey</p>
    <p>https://codepen.io/random-page</p>
    <p>https://blog.codepen.io</p>
    <p>https://blog.codepen.io/user/embed/123456</p>
    <p>https://blog.codepen.io/user/pen/123456</p>
    <p>https://codepen.io/team/codepen/embed/PNaGbb</p>
    <p>https://codepen.io/team/codepen/embed/PNaGbb?default-tab=js</p>
    <iframe src="https://codepen.io/team/codepen/embed/preview/PNaGbb" style="width:100%; height:300px;"></iframe>
    <iframe src="https://www.codepen.io/team/codepen/embed/preview/PNaGbb" style="width:100%; height:300px;"></iframe>
    <iframe src="https://codepen.io/chriscoyier/embed/preview/owBwKM" style="width:100%; height:300px;"></iframe>
    <iframe src="https://www.codepen.io/chriscoyier/embed/preview/owBwKM" style="width:100%; height:300px;"></iframe>

  `);
});
