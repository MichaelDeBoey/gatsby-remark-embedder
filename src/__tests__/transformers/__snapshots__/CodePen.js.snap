// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can transform CodePen links 1`] = `
"<iframe src=\\"https://codepen.io/team/codepen/embed/preview/PNaGbb\\" style=\\"width:100%; height:300px;\\"></iframe>
"
`;
